Sources/TuistAsyncQueue/AsyncQueuePersistor.swift

```diff
@@ -0,0 +1,108 @@
+import Foundation
+import RxSwift
+import TSCBasic
+import TuistCore
+import TuistSupport
+
+typealias AsyncQueueEventTuple = (dispatcherId: String, id: UUID, date: Date, data: Data, filename: String)
+
+protocol AsyncQueuePersisting {
+    /// Reads all the persisted events and returns them.
+    func readAll() -> Single<[AsyncQueueEventTuple]>
+
+    /// Persiss a given event.
+    /// - Parameter event: Event to be persisted.
+    func write<T: AsyncQueueEvent>(event: T) -> Completable
+
+    /// Deletes the given event from disk.
+    /// - Parameter event: Event to be deleted.
+    func delete<T: AsyncQueueEvent>(event: T) -> Completable
+
+    /// Deletes the given file name from disk.
+    /// - Parameter filename: Name of the file to be deleted.
+    func delete(filename: String) -> Completable
+}
+
+final class AsyncQueuePersistor: AsyncQueuePersisting {
+    // MARK: - Attributes
+
+    let directory: AbsolutePath
+    let jsonEncoder: JSONEncoder = JSONEncoder()
+
+    // MARK: - Init
+
+    init(directory: AbsolutePath = Environment.shared.queueDirectory) {
+        self.directory = directory
+    }
+
+    func write<T: AsyncQueueEvent>(event: T) -> Completable {
+        Completable.create { (observer) -> Disposable in
+            let path = self.directory.appending(component: self.filename(event: event))
+            do {
+                let data = try self.jsonEncoder.encode(event)
+                try data.write(to: path.url)
+                observer(.completed)
+            } catch {
+                observer(.error(error))
+            }
+            return Disposables.create()
+        }
+    }
+
+    func delete<T: AsyncQueueEvent>(event: T) -> Completable {
+        delete(filename: filename(event: event))
+    }
+
+    func delete(filename: String) -> Completable {
+        Completable.create { (observer) -> Disposable in
+            let path = self.directory.appending(component: filename)
+            guard FileHandler.shared.exists(path) else { return Disposables.create() }
+            do {
+                try FileHandler.shared.delete(path)
+                observer(.completed)
+            } catch {
+                observer(.error(error))
+            }
+            return Disposables.create()
+        }
+    }
+
+    func readAll() -> Single<[AsyncQueueEventTuple]> {
```

Comment:
@RomainBoulay what do you think about adding a bit of documentation somewhere in this file where we describe the convention that we follow for persisting the events. It'll make it easier for developers in the future to understand the logic in this method.